### PR TITLE
Don't pin package:macros

### PR DESCRIPTION
**Pull request by christopherfujino**
Full context https://github.com/flutter/flutter/issues/148004

TL;DR

1. Because `package:macros` depends on `package:_macros` via [a Dart SDK dependency](https://github.com/dart-lang/sdk/blob/main/pkg/macros/pubspec.yaml#L13); and
2. `package:macros` exactly pins `package:_macros`; and
3. Each new version of `package:macros` will increment its exact pin of `package:_macros` ([comment](https://github.com/flutter/flutter/issues/148004#issuecomment-2103099893)); and
4. The [flutter/flutter](https://github.com/flutter/flutter) repository exactly pins the Dart SDK (via its exact pin on the [flutter/engine](https://github.com/flutter/engine) repository

Therefore, the [flutter/flutter](https://github.com/flutter/flutter) repository effectively pins both `package:macros` and `package:_macros` already (as in, there exists only a single version of each that pub will successfully be able to solve within the context of a particular Flutter SDK).

Therefore, it is safe for [flutter/flutter](https://github.com/flutter/flutter) repository to *not* pin `package:macros`, which will allow engine -> framework rolls that contain a new Dart SDK that contains a new `package:_macros` version to land automatically, provided all tests pass.

### CODE REVIEW

**Comment by christopherfujino** [comment]
For science, I pushed the commit https://github.com/flutter/flutter/pull/148087/commits/58a3acdceaea64b686e524bddbc2ee44c66237d3, which rolls the engine to an older roll that failed CI in https://github.com/flutter/flutter/pull/148008, because pub was not able to solve successfully with the exact pins of `package:macros`. If CI passes on that commit, I will revert it and land this change without it, and then re-enable the dart -> engine roller.

**Comment by christopherfujino** [comment]
> For science, I pushed the commit [58a3acd](https://github.com/flutter/flutter/commit/58a3acdceaea64b686e524bddbc2ee44c66237d3), which rolls the engine to an older roll that failed CI in #148008, because pub was not able to solve successfully with the exact pins of `package:macros`. If CI passes on that commit, I will revert it and land this change without it, and then re-enable the dart -> engine roller (FYI @rmacnak-google)

All presubmits passed except for two integration test flakes unrelated to pub solving `package:macros`, so I will call this a successful test: https://github.com/flutter/flutter/pull/148087/checks?check_run_id=24794486500

**Review by zanderso — APPROVED** [review]

**Comment by christopherfujino** [comment]
Merging manually because autosubmit bot is broken: https://github.com/flutter/flutter/issues/148092
